File: problems/C++/64. Minimum Path Sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        
        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                if (r == 0 && c == 0) continue;
                if (r == 0) {
                    grid[r][c] += grid[r][c - 1];
                } else if (c == 0) {
                    grid[r][c] += grid[r - 1][c];
                } else {
                    grid[r][c] += min(grid[r - 1][c], grid[r][c - 1]);
                }
            }
        }
        return grid[m - 1][n - 1];
    }
};

File: problems/Java/64. Minimum Path Sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        
        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                if (r == 0 && c == 0) continue;
                
                int up = (r > 0) ? grid[r - 1][c] : Integer.MAX_VALUE;
                int left = (c > 0) ? grid[r][c - 1] : Integer.MAX_VALUE;
                
                if (up != Integer.MAX_VALUE || left != Integer.MAX_VALUE) {
                    grid[r][c] += Math.min(up, left);
                }
            }
        }
        return grid[m - 1][n - 1];
    }
}